File: lab/mar_14/solution.cc
/*
* author: jamie
* email: jinjiedeng.jjd@gmail.com
*/
#include <deque>
#include <iostream>
#include <vector>

int main ()
{
  uint64_t n, m;
  std::cin >> n >> m;
  std::vector<std::deque<uint64_t>> barrels(n + 1);
  while (m--)
  {
    uint64_t op, i, j, k;
    std::cin >> op;
    if (op) // retrieve
    {
      std::cin >> j >> k;
      if (j) // the-1-end
      {
        std::cout << barrels[k].back() << "\n";
        barrels[k].pop_back();
      }
      else // the-0-end
      {
        std::cout << barrels[k].front() << "\n";
        barrels[k].pop_front();
      }
    }
    else // store
    {
      std::cin >> i >> j >> k;
      if (j) { barrels[k].push_back(i); } // the-1-end
      else   { barrels[k].push_front(i); } // the-0-end
    }
  }
  return 0;
}

/*
* author: redleaf
* email: redleaf23477@gapp.nthu.edu.tw
*/
#include <bits/stdc++.h>
using namespace std;
using LL = long long int;

int main() {
    int n; cin >> n;
    int m; cin >> m;
    vector<deque<int>> deq(n);
    while (m--) {
        int op; cin >> op;
        if (op == 0) {
            int i, j, k; cin >> i >> j >> k;
            if (j == 0) deq[k-1].emplace_front(i);
            else deq[k-1].emplace_back(i);
        } else {
            int j, k; cin >> j >> k;
            if (j == 0) {
                cout << deq[k-1].front() << "\n";
                deq[k-1].pop_front();
            } else {
                cout << deq[k-1].back() << "\n";
                deq[k-1].pop_back();
            }
        }
    }
}

File: lab/mar_14/solution.py
#
# author: wang-yang
# email: tnst92002@gmail.com
#
from collections import deque

n = int(input())
m = int(input())

dq = [deque() for _ in range(n+1)]
for _ in range(m):
    op, i, j, *k = map(int, input().split())
    k = k[0] if k else None
    if op == 0:
        if j == 0:
            dq[k].append(i)
        else:
            dq[k].appendleft(i)
    else:
        if i == 0:
            val = dq[j].pop()
        else:
            val = dq[j].popleft()
        print(val)

#
# author: redleaf
# email: redleaf23477@gapp.nthu.edu.tw
#
from collections import deque

n = int(input())
m = int(input())

deq = [deque() for _ in range(n)]

for _ in range(m):
    ops = list(map(int, input().split()))
    if ops[0] == 0:
        op, i, j, k = ops
        if j == 0:
            deq[k-1].appendleft(i)
        else:
            deq[k-1].append(i)
    else:
        op, j, k = ops
        if j == 0:
            print(deq[k-1].popleft())
        else:
            print(deq[k-1].pop())
